presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ColumnRangesSystemTable.java

```diff
@@ -0,0 +1,175 @@
+                .listTableColumns(sourceTable.getTableId()).stream()
+                .filter(column -> isIndexedType(column.getDataType()))
+                .collect(toImmutableList());
+        List<ColumnMetadata> systemTableColumns = indexedRaptorColumns.stream()
+                .flatMap(column -> Stream.of(
+                        new ColumnMetadata(column.getColumnName() + MIN_COLUMN_SUFFIX, column.getDataType(), null, false),
+                        new ColumnMetadata(column.getColumnName() + MAX_COLUMN_SUFFIX, column.getDataType(), null, false)))
+                .collect(toImmutableList());
+        SchemaTableName tableName = new SchemaTableName(sourceTable.getSchemaName(), sourceTable.getTableName() + COLUMN_RANGES_TABLE_SUFFIX);
+        this.tableMetadata = new ConnectorTableMetadata(tableName, systemTableColumns);
+    }
+
+    public static Optional<SchemaTableName> getSourceTable(SchemaTableName tableName)
+    {
+        if (tableName.getTableName().endsWith(COLUMN_RANGES_TABLE_SUFFIX) &&
+                !tableName.getTableName().equals(COLUMN_RANGES_TABLE_SUFFIX)) {
+            int tableNameLength = tableName.getTableName().length() - COLUMN_RANGES_TABLE_SUFFIX.length();
+            return Optional.of(new SchemaTableName(
+                    tableName.getSchemaName(),
+                    tableName.getTableName().substring(0, tableNameLength)));
+        }
+        return Optional.empty();
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        String metadataSqlQuery = getColumnRangesMetadataSqlQuery(sourceTable, indexedRaptorColumns);
+        List<Type> columnTypes = tableMetadata.getColumns().stream()
+                .map(ColumnMetadata::getType)
+                .collect(toImmutableList());
+
+        PageListBuilder pageListBuilder = new PageListBuilder(columnTypes);
+
+        try (Connection connection = dbi.open().getConnection();
+                Statement statement = connection.createStatement();
+                ResultSet resultSet = statement.executeQuery(metadataSqlQuery)) {
+            if (resultSet.next()) {
+                pageListBuilder.beginRow();
+                for (int i = 0; i < columnTypes.size(); ++i) {
+                    Type columnType = columnTypes.get(i);
+                    if (columnType.equals(BIGINT) || columnType.equals(DATE) || columnType.equals(TIMESTAMP)) {
+                        long value = resultSet.getLong(i + 1);
+                        if (!resultSet.wasNull()) {
+                            columnType.writeLong(pageListBuilder.nextBlockBuilder(), value);
```

Comment:
Call `pageListBuilder.nextBlockBuilder()` at the top of the loop. This makes it easier to verify that we are calling it exactly once for each column.